crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -0,0 +1,224 @@
+        + Storage
+        + ParentSync
+        + ChildObjectResolver,
+> {
+    vm: &'v MoveVM,
+    state_view: &'s S,
+    ctx: &'a mut TxContext,
+    gas_status: &'a mut GasStatus<'b>,
+    session: Option<Session<'s, 'v, S>>,
+    object_owner_map: BTreeMap<ObjectID, Owner>,
+    pub gas: Option<Value>,
+    pub inputs: Vec<Option<Value>>,
+    pub results: Vec<Vec<Option<Value>>>,
+    // transfers not from the Move runtime
+    additional_transfers: Vec<(/* new owner */ SuiAddress, ObjectValue)>,
+}
+impl<'v, 's, 'a, 'b, E, S> ExecutionContext<'v, 's, 'a, 'b, E, S>
+where
+    E: fmt::Debug,
+    S: ResourceResolver<Error = E>
+        + ModuleResolver<Error = E>
+        + Storage
+        + ParentSync
+        + ChildObjectResolver,
+{
+    pub fn new(
+        vm: &'v MoveVM,
+        state_view: &'s S,
+        ctx: &'a mut TxContext,
+        gas_status: &'a mut GasStatus<'b>,
+        gas_coin: ObjectID,
+        inputs: Vec<CallArg>,
+    ) -> Result<Self, ExecutionError> {
+        let mut object_owner_map = BTreeMap::new();
+        let inputs = inputs
+            .into_iter()
+            .map(|call_arg| {
+                Ok(Some(load_call_arg(
+                    state_view,
+                    &mut object_owner_map,
+                    call_arg,
+                )?))
+            })
+            .collect::<Result<_, ExecutionError>>()?;
+        let gas = Some(Value::Object(load_object(
+            state_view,
+            &mut object_owner_map,
+            gas_coin,
+        )?));
+        Ok(Self {
+            vm,
+            state_view,
+            ctx,
+            gas_status,
+            session: None,
+            object_owner_map,
+            gas,
+            inputs,
+            results: vec![],
+            additional_transfers: vec![],
+        })
+    }
+
+    pub fn session(&mut self) -> &Session<'s, 'v, S> {
+        self.session.get_or_insert_with(|| {
+            new_session(
+                self.vm,
+                self.state_view,
+                self.object_owner_map.clone(),
+                self.gas_status.is_metered(),
+            )
+        })
+    }
+
+    pub fn fresh_id(&mut self) -> Result<ObjectID, ExecutionError> {
+        if true {
+            todo!("update native context set")
+        }
+        Ok(self.ctx.fresh_id())
+    }
+
+    pub fn delete_id(&mut self, _object_id: ObjectID) -> Result<(), ExecutionError> {
+        if true {
+            todo!("update native context set")
+        }
+        Ok(())
+    }
+
+    pub fn take_args<V: TryFromValue>(
+        &mut self,
+        args: Vec<Argument>,
+    ) -> Result<Vec<V>, ExecutionError> {
+        args.into_iter()
+            .map(|arg| self.take_arg::<V>(arg))
+            .collect()
+    }
+
+    pub fn take_arg<V: TryFromValue>(&mut self, arg: Argument) -> Result<V, ExecutionError> {
+        let val_opt = self.borrow_mut(arg)?;
+        if val_opt.is_none() {
+            panic!("taken value")
```

Comment:
I've used panics instead of `ExecutionErrors` as I will need to stamp out new errors for these, and I want to wait until I'm confident that I have all the errors I want/need :)